src/hooks/useDirectMessageEndCursor.tsx

```diff
@@ -0,0 +1,81 @@
+import { useEffect, useState } from 'react';
+import { usePrivatePostsForUsers } from './Circles/usePrivatePosts';
+import { useAsyncStorage } from './useAsyncStorage';
+import AsyncStorage from '@react-native-async-storage/async-storage';
+import { isEqual } from 'lodash';
+
+export const useDirectMessageEndCursor = (
+  userId: string,
+  recipientUserId: string,
+) => {
+  const [endCursor, setEndCursor] = useAsyncStorage(
+    `${userId}-${recipientUserId}-endCursor`,
+  );
+
+  return { endCursor: endCursor?.data, setEndCursor };
+};
+
+type Props = {
+  currentUserId?: string;
+  userIds: string[];
+};
+
+const useGetLocalCursorsForUsers = ({ currentUserId, userIds }: Props) => {
+  const [result, setResult] = useState<Map<string, string>>();
+  useEffect(() => {
+    async function fetchEndCursors() {
+      const endCursorMap = new Map<string, string>();
+
+      for await (const userId of userIds) {
+        const endCursor = await AsyncStorage.getItem(
+          `${currentUserId}-${userId}-endCursor`,
+        );
+        if (endCursor) {
+          endCursorMap.set(userId, endCursor);
+        }
+      }
+      if (!isEqual(result, endCursorMap)) {
+        setResult(endCursorMap);
+      }
+    }
+    fetchEndCursors();
+  });
+
+  return result;
+};
+
+const useGetRemoteCursorsForUsers = ({ userIds }: Props) => {
+  const endCursorMap = new Map<string, string>();
+  const userQueries = usePrivatePostsForUsers(userIds);
+  userQueries.forEach(async (userQuery, index) => {
```

Comment:
This `forEach(async` looks like it might not actually await them each blocks.  I'd recommend switching to `for .. of` here.